File: scaffold/src/main/java/com/vgaw/scaffold/view/ImgTxtLayout.java
package com.vgaw.scaffold.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.vgaw.scaffold.R;
import com.vgaw.scaffold.util.phone.DensityUtil;

public class ImgTxtLayout extends LinearLayout {
    protected Drawable mIcon;
    protected Drawable mIconBackground;
    protected int mIconSize;
    protected String mName;
    protected int mNameColor;
    protected int mNameSize;
    protected int mPadding;

    private ImageView mImgTxtIv;
    private TextView mImgTxtTv;

    public ImgTxtLayout(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public ImgTxtLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public ImgTxtLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(attrs);
    }

    protected void setIcon(Drawable icon) {
        mImgTxtIv.setImageDrawable(icon);
    }

    public void setContent(String content) {
        mImgTxtTv.setText(content);
    }

    private void init(AttributeSet attrs) {
        TypedArray array = getContext().obtainStyledAttributes(attrs, R.styleable.ImgTxtLayout);
        mIcon = array.getDrawable(R.styleable.ImgTxtLayout_imgTxtIcon);
        mIconBackground = array.getDrawable(R.styleable.ImgTxtLayout_imgTxtIconBackground);
        mIconSize = array.getDimensionPixelSize(R.styleable.ImgTxtLayout_imgTxtIconSize, DensityUtil.dp2px(getContext(), 48));
        mName = array.getString(R.styleable.ImgTxtLayout_imgTxtName);
        mNameColor = array.getColor(R.styleable.ImgTxtLayout_imgTxtNameColor, getResources().getColor(R.color.white));
        mNameSize = array.getDimensionPixelSize(R.styleable.ImgTxtLayout_imgTxtNameSize, DensityUtil.sp2px(getContext(), 16));
        mPadding = array.getDimensionPixelSize(R.styleable.ImgTxtLayout_imgTxtPadding, DensityUtil.sp2px(getContext(), 4));

        array.recycle();

        refreshChildParam();

        mImgTxtIv = new ImageView(getContext());
        mImgTxtIv.setScaleType(ImageView.ScaleType.CENTER_CROP);
        mImgTxtIv.setImageDrawable(mIcon);
        mImgTxtIv.setBackground(mIconBackground);
        addView(mImgTxtIv, proGravity(new LayoutParams(mIconSize, mIconSize)));

        mImgTxtTv = new TextView(getContext());
        mImgTxtTv.setText(mName);
        mImgTxtTv.setTextColor(mNameColor);
        mImgTxtTv.setTextSize(TypedValue.COMPLEX_UNIT_PX, mNameSize);
        LayoutParams layoutParams = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        proPadding(layoutParams);
        addView(mImgTxtTv, proGravity(layoutParams));
    }

    private LayoutParams proPadding(LayoutParams params) {
        int orientation = getOrientation();
        if (orientation == HORIZONTAL) {
            params.leftMargin = mPadding;
        } else {
            params.topMargin = mPadding;
        }
        return params;
    }

    private LayoutParams proGravity(LayoutParams params) {
        int orientation = getOrientation();
        if (orientation == HORIZONTAL) {
            params.gravity = Gravity.CENTER_VERTICAL;
        } else {
            params.gravity = Gravity.CENTER_HORIZONTAL;
        }
        return params;
    }

    protected void refreshChildParam() {}
}

File: scaffold/src/main/java/com/vgaw/scaffold/view/ImgTxtLayout.kt
package com.vgaw.scaffold.view;

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.ColorInt
import com.vgaw.scaffold.R
import com.vgaw.scaffold.util.Util
import com.vgaw.scaffold.util.phone.DensityUtil

open class ImgTxtLayout : LinearLayout {
    protected var mIcon: Drawable? = null
    protected var mIconBackground: Drawable? = null
    protected var mIconSize: Int = -1
    protected var mName: String? = null
    protected var mNameColor: Int = -1
    protected var mNameSize: Int = -1
    protected var mPadding: Int = -1

    private lateinit var mImgTxtIv: ImageView
    private lateinit var mImgTxtTv: TextView

    constructor(context: Context, attrs: AttributeSet): super(context, attrs) {
        init(attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
        init(attrs)
    }

    fun setIcon(icon: Drawable?) {
        mImgTxtIv.setImageDrawable(icon)
    }

    fun setContent(content: String?) {
        mImgTxtTv.text = Util.nullToEmpty(content)
    }

    fun setContentColor(@ColorInt color: Int) {
        mImgTxtTv.setTextColor(color)
    }

    protected fun refreshChildParam() {}

    private fun init(attrs: AttributeSet) {
        val array = getContext().obtainStyledAttributes(attrs, R.styleable.ImgTxtLayout)
        val ignoreSelectWave = array.getBoolean(R.styleable.ImgTxtLayout_imgTxtIgnoreSelectWave, false)
        if (!ignoreSelectWave) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                foreground = resources.getDrawable(R.drawable.selectable_item_bg)
            }
        }
        mIcon = array.getDrawable(R.styleable.ImgTxtLayout_imgTxtIcon)
        mIconBackground = array.getDrawable(R.styleable.ImgTxtLayout_imgTxtIconBackground)
        mIconSize = array.getDimensionPixelSize(R.styleable.ImgTxtLayout_imgTxtIconSize, DensityUtil.dp2px(getContext(), 48F))
        mName = array.getString(R.styleable.ImgTxtLayout_imgTxtName)
        mNameColor = array.getColor(R.styleable.ImgTxtLayout_imgTxtNameColor, getResources().getColor(R.color.white))
        mNameSize = array.getDimensionPixelSize(R.styleable.ImgTxtLayout_imgTxtNameSize, DensityUtil.sp2px(getContext(), 16F))
        mPadding = array.getDimensionPixelSize(R.styleable.ImgTxtLayout_imgTxtPadding, DensityUtil.sp2px(getContext(), 4F))

        array.recycle()

        refreshChildParam()

        mImgTxtIv = ImageView(context)
        mImgTxtIv.scaleType = ImageView.ScaleType.CENTER_CROP
        mImgTxtIv.setImageDrawable(mIcon)
        mImgTxtIv.background = mIconBackground
        addView(mImgTxtIv, proGravity(LayoutParams(mIconSize, mIconSize)))

        mImgTxtTv = TextView(context)
        mImgTxtTv.text = mName
        mImgTxtTv.setTextColor(mNameColor)
        mImgTxtTv.setTextSize(TypedValue.COMPLEX_UNIT_PX, mNameSize.toFloat())
        val layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        proPadding(layoutParams)
        addView(mImgTxtTv, proGravity(layoutParams))
    }

    private fun proPadding(params: LayoutParams): LayoutParams {
        if (orientation == HORIZONTAL) {
            params.leftMargin = mPadding
        } else {
            params.topMargin = mPadding
        }
        return params
    }

    private fun proGravity(params: LayoutParams): LayoutParams {
        if (orientation == HORIZONTAL) {
            params.gravity = Gravity.CENTER_VERTICAL
        } else {
            params.gravity = Gravity.CENTER_HORIZONTAL
        }
        return params
    }
}
